### ISSUE TITLE
dhcp: patch up static leases with missing prefix et al

### ISSUE THREAD

**Issue by fichtner** [opnsense/core#4642]
**Important notices**

Before you add a new report, we ask you kindly to acknowledge the following:

- [x] I have read the contributing guide lines at https://github.com/opnsense/core/blob/master/CONTRIBUTING.md
- [x] I have searched the existing issues and I am convinced that mine is new.

**Describe the bug**

When tracking DHCPv6 manually you can set suffix addresses and DHCPv6/RADVD completes them, but Unbound and Dnsmasq do not.

While here it would make sense to create an iterator to get rid of all the duplication in the code for IPv4 and IPv6 staticmap handling.

**To Reproduce**

See forum link.

**Expected behavior**

Patch up addresses in the manual tracking case but do not solve the changing prefix issue.

**Screenshots**

N/A

**Relevant log files**

N/A

**Additional context**

https://forum.opnsense.org/index.php?topic=21108.0
https://github.com/opnsense/core/issues/2544

**Environment**

Most.

**Comment by maurice-w** [comment]
Isn't that a duplicate of #3657? Don't get me wrong, I would appreciate a solution, but am a bit surprised about the sudden change of mind.

**Comment by fichtner** [comment]
Yes, looks like it. I switched jobs in January to work on OPNsense full time so this is more of a reachable scope now.

**Comment by fichtner** [comment]
PR at https://github.com/opnsense/core/pull/4723

**Comment by maurice-w** [comment]
d0822b006785be56d3cad925e5f5f04c63de1d9e + e73db9c7aef5cfaf63e1c23637dafa29885ae58b seems to fix it for DHCPv6 + Unbound. 👍

Didn't test dnsmasq, didn't test DHCPv4.

Following up on my comment at d0822b006785be56d3cad925e5f5f04c63de1d9e:

Would be nice to have the full addresses in status_dhcpv6_leases.php, too. Not a functional issue, just a cosmetic one.

While at it, can we avoid restarting radvd every time the dhcpd6 config changes? It's a long standing issue which causes the entire network to go down briefly when you e.g. add a static mapping. I guess we'd have to find out whether this entanglement is actually required or purely historic.

**Comment by fichtner** [comment]
Hmm, the code was pretty bad, it should have checked for _ipaddrv6_:

https://github.com/opnsense/core/commit/db29e02ecbfded6298da2c01b444a3ce01ebd264#diff-892d19c546842230914680258f3f81046a0d541c8f0196bd82a1dbe628c3c82bL461

Anyway, db29e02 should be an improvement on the leases page. That's half of your requests done :)

**Comment by maurice-w** [comment]
Did a quick test and it's looking good, thanks!

**Comment by fichtner** [comment]
We could pull the radvd reconfiguration from the main restart sequence and start it one layer up manually after dhcp reconfiguration. The resulting commit would be enough to discuss which calls need to stay and which need to be removed, ok?

**Comment by marjohn56** [comment]
@fichtner - please join our discussion on #4768

**Comment by maurice-w** [comment]
@fichtner Sounds good to me!

**Comment by fichtner** [comment]
@maurice-w can you have a look at 012e0f7f11f0f62 ? Slightly different approach, see commit message

**Comment by maurice-w** [comment]
Definitely an improvement. Changing DHCP settings doesn't restart radvd anymore. 👍

The only side effect I've noticed so far: If _Use the DNS settings of the DHCPv6 server_ is enabled in the RA settings, changing the DHCPv6 DNS servers now requires a manual radvd restart to update RDNSS in radvd.conf. No surprise.
Applying Unbound configuration changes also triggers a radvd restart - probably because of possible RDNSS changes, too. Maybe we could restart radvd only if RDNSS actually changes?

**Comment by fichtner** [comment]
There is a likely issue with the v6 leases page now expecting an IP address and a hostname so static leases may not show up. The question is what should be mandatory for static leases and what can we fill out  from the lease page add button.

IPv6 also not merging system leases and static leases. What is their respective lookup key.. the IP address or DUID or a combination?

Reference: https://forum.opnsense.org/index.php?topic=22742.0

**Comment by maurice-w** [comment]
Hm, the bug report on the forum is a bit hard to follow...
Static mappings without a hostname don't show up on the leases page? Possible, I never tested this.

> The question is what should be mandatory for static leases and what can we fill out from the lease page add button.

Technically, a static mapping only requires a DUID and an IP address. Some DHCPv6 servers (e.g. Microsoft) won't allow reservations without a hostname though, so making this mandatory might be acceptable. For converting dynamic leases to static mappings, we should grab DUID + IP address + hostname from the leases page. Isn't that the case? Never really used that feature.

> IPv6 also not merging system leases and static leases. What is their respective lookup key.. the IP address or DUID or a combination?

Sorry, I don't understand. Could you elaborate? Unless this is just a "note to self". :-)

**Comment by fichtner** [comment]
All code for this issue is now in 21.1.6 and we are almost there. One small regression reported with Unbound waiting for confirmation:

https://forum.opnsense.org/index.php?topic=23304.0

**Comment by fichtner** [comment]
> > IPv6 also not merging system leases and static leases. What is their respective lookup key.. the IP address or DUID or a combination?
> Sorry, I don't understand. Could you elaborate? Unless this is just a "note to self". :-)

yes, IPv6 leases page is a little weird still... otherwise I think we are set for 21.7.

**Comment by fichtner** [comment]
@maurice-w I'm following your suggestion to make the IP mandatory for static lease consumers. For DHCP registration in DNS obviously we need to double-check for the hostname otherwise there is nothing to register. DHCPv6 leases page now merges static entries on DUID match and the DHCPv4 leases page will also use dhcpd_staticmap() which brings this particular saga to a happy end.

@marjohn56 @maurice-w Thanks for all the help!


Cheers,
Franco
